keeper/upkeep_executer.go

```diff
@@ -0,0 +1,280 @@
+	return upkeepExecuter{
+		blockHeight:    atomic.NewUint64(0),
+		chainlinkNode:  clNode,
+		endpoint:       config.KeeperEthEndpoint,
+		registryStore:  NewRegistryStore(dbClient, uint64(config.KeeperBlockCooldown)),
+		executionQueue: make(chan struct{}, executionQueueSize),
+		chDone:         make(chan struct{}),
+		chSignalRun:    make(chan struct{}, 1),
+	}
+}
+
+type upkeepExecuter struct {
+	blockHeight   *atomic.Uint64
+	chainlinkNode chainlink.Node
+	endpoint      string
+	ethClient     *ethclient.Client
+	registryStore RegistryStore
+
+	executionQueue chan struct{}
+	chDone         chan struct{}
+	chSignalRun    chan struct{}
+}
+
+var _ UpkeepExecuter = upkeepExecuter{} // upkeepExecuter satisfies UpkeepExecuter
+
+func (executer upkeepExecuter) Start() error {
+	ethClient, err := ethclient.Dial(executer.endpoint)
+	if err != nil {
+		return err
+	}
+	executer.ethClient = ethClient
+
+	if strings.HasPrefix(executer.endpoint, "ws") {
+		go executer.setRunsOnHeadSubscription()
+	} else if strings.HasPrefix(executer.endpoint, "http") {
+		go executer.setRunsOnTimer()
+	} else {
+		return fmt.Errorf("unknown endpoint protocol: %+v", executer.endpoint)
+	}
+
+	go executer.run()
+
+	return nil
+}
+
+func (executer upkeepExecuter) Stop() {
+	close(executer.chDone)
+}
+
+func (executer upkeepExecuter) run() {
+	for {
+		select {
+		case <-executer.chDone:
+			return
+		case <-executer.chSignalRun:
+			executer.processActiveRegistrations()
+		}
+	}
+}
+
+func (executer upkeepExecuter) processActiveRegistrations() {
+	// TODO - RYAN - this should be batched to avoid congestgion
+	logger.Debug("received new block, running checkUpkeep for keeper registrations")
+
+	activeRegistrations, err := executer.registryStore.Active(executer.blockHeight.Load())
+	if err != nil {
+		logger.Errorf("unable to load active registrations: %v", err)
+		return
+	}
+
+	for _, reg := range activeRegistrations {
+		executer.concurrentExecute(reg)
+	}
+}
+
+func (executer upkeepExecuter) concurrentExecute(registration registration) {
+	executer.executionQueue <- struct{}{}
+	go executer.execute(registration)
+}
+
+// execute will call checkForUpkeep and, if it succeeds, triger a job on the CL node
+func (executer upkeepExecuter) execute(registration registration) {
+	// pop queue when done executing
+	defer func() {
+		<-executer.executionQueue
+	}()
+
+	checkPayload, err := upkeepRegistryABI.Pack(
+		checkUpkeep,
+		big.NewInt(int64(registration.UpkeepID)),
+		registration.Registry.From,
+	)
+	if err != nil {
+		logger.Error(err)
+		return
+	}
+
+	msg := ethereum.CallMsg{
+		From: utils.ZeroAddress,
+		To:   &registration.Registry.Address,
+		Gas:  20000000, // TODO - RYAN - https://www.pivotaltracker.com/story/show/176391267
+		Data: checkPayload,
+	}
+
+	result, err := executer.ethClient.CallContract(context.Background(), msg, nil)
+	if err != nil {
+		// don't log anything as this is extremely common and would be too noisey
```

Comment:
Maybe log something so we know? Is it possible to check for the revert message?

Reply:
as far as I can tell, the revert message is always the generic "EVM revert... " I might be missing something though. do you know if geth usually pulls the revert message from the tx receipt?